### PR TITLE
DAOS-16355 pydaos: add dir object cache for Datasets

### PR DESCRIPTION
**Pull request by 0xE0F**
Dataset reads a lot of samples that are stored under just a few common directories, to reduce computation load on lookup for each sample file, this commit introduces a cache of directory objects.

Features: PytorchDatasetsTest,PytorchCheckpointTest

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

**Review by johannlombardi — DISMISSED** [review]

**Review by jolivier23 — DISMISSED** [review]

File: src/client/pydaos/torch/torch_shim.c [hunk 1]
```diff
@@ -95,6 +152,50 @@ __shim_handle__module_fini(PyObject *self, PyObject *args)
 	return PyLong_FromLong(rc);
 }
 
+/* Hash table does not require lock: workers will be working on their own subsets of data
+and will have their own handlers, including the cache.
+*/
+static int
+__dir_cache_create(struct dfs_handle *hdl)
+{
+	int rc = d_hash_table_create(D_HASH_FT_EPHEMERAL | D_HASH_FT_NOLOCK | D_HASH_FT_LRU, 4,
```
Comment (mchaarawi) [inline src/client/pydaos/torch/torch_shim.c:161]: 4 means 2^4 (16) entries allowed in the hashtable. i thought for pytorch workloads, the number of dirs is usually much greater than that. do you want to reconsider this?
If you have done perf measurement where the number of directories in your dataset is greater than that, then i doubt you have seen any improvement in perf.

**Review by mchaarawi — CHANGES_REQUESTED** [review]
no other changes requested other than verifying the size of the hashtable is sufficient.

File: src/client/pydaos/torch/torch_shim.c (hunk 1 above)
Comment (enakta) [reply src/client/pydaos/torch/torch_shim.c:161]: I tested with two labels dataset: `/samples/train/{0,1}/<samples>`, which is default (and unet3d) layout for dlio_benchmark.
I don't think there would be one size that fits them all and the parameter in the constructor would be more flexible approach, indeed.

Comment (enakta) [reply src/client/pydaos/torch/torch_shim.c]: @mchaarawi added ability to passthrough cache size from the client code

Comment (mchaarawi) [reply src/client/pydaos/torch/torch_shim.c]: sorry i should have been more clear.
for the change that you just made, 2^256 for the dir cache size is too extreme and likely consumes too much memory. i would suggest to use something like 16, which should be ok for the default, or even for always.

in the example you posted, are all the datasets you tested with 2 directories under train? and all the files under those 2 directories? those should have fit in the cache and you should decrease the number of lookups. I have seen datasets that have thousands of directories (imagenet after being transformed, etc).  so just wanted to confirm.

Comment (enakta) [reply src/client/pydaos/torch/torch_shim.c]: > in the example you posted, are all the datasets you tested with 2 directories under train? and all the files under those 2 directories? those should have fit in the cache and you should decrease the number of lookups. 

Yes, it did. 

@mchaarawi it's not 2^256 tho,  there's conversion  with `uint32_t bits = ceil(log2(hdl->dir_cache_size))`.
I just didn't want to expose it to client code as 2^N to avoid confusion 😬

**Review by mchaarawi — APPROVED** [review]

File: src/client/pydaos/torch/torch_shim.c [hunk 2]
```diff
@@ -262,32 +383,78 @@ __shim_handle__torch_worker_init(PyObject *self, PyObject *args)
...
+	}
+
+	rec->obj = *obj;
+	strncpy(rec->name, name, len);
+
+	rc = d_hash_rec_insert(hdl->dir_cache, rec->name, len, &rec->entry, false);
+	if (rc) {
+		D_ERROR("Failed to insert dir handle in hashtable: '%s': %s (rc=%d)", name,
+			d_errstr(rc), rc);
+		rc = daos_der2errno(rc);
```
Comment (daltonbohning) [inline src/client/pydaos/torch/torch_shim.c:426]: Should we instead print this as a system errno instead of d_err? I.e. convert der2errno and then print

**Review by daltonbohning — CHANGES_REQUESTED** [review]
Code LGTM but this
```
Features: PytorchDatasetsTest,PytorchCheckpointTest
```
Means "run tests tagged with both PytorchDatasetsTest and PytorchCheckpointTest"
What you want is
```
Features: PytorchDatasetsTest PytorchCheckpointTest
```
Or, more simply:
```
Features: pytorch
```

File: src/client/pydaos/torch/torch_shim.c (hunk 2 above)
Comment (enakta) [reply src/client/pydaos/torch/torch_shim.c:426]: `d_errstr` might give more specific details in case of the error, rather than more generic `errno`

Comment (enakta) [reply src/client/pydaos/torch/torch_shim.c:426]: @daltonbohning thanks, I've restarted checks with `Features: pytorch`

Comment (daltonbohning) [reply src/client/pydaos/torch/torch_shim.c:426]: I mean since all other printed errors at this layer are errno, it seems inconsistent to print d_err here

**Review by daltonbohning — APPROVED** [review]

File: src/client/pydaos/torch/torch_shim.c (hunk 2 above)
Comment (enakta) [reply src/client/pydaos/torch/torch_shim.c:426]: ah I see, but there are heaps of other `d_errstr` in this layer: pretty much everything that does not start with `dfs_` prefix is printed with `d_errstr` and then converted to `errno` with `daos_der2errno`. Just to have a bit more detail when an error pops up - saved me a fair few grey hairs.
